### DISCUSSION TITLE
I wrote a Tree-sitter grammar for Djot

### DISCUSSION THREAD

**Discussion by treeman** [jgm/djot#280] [Show and tell]
Things happened, and I ended up creating a Tree-sitter grammar for Djot: https://github.com/treeman/tree-sitter-djot

It should be feature complete with Djot as described by the syntax description, but I'd appreciate any bug reports and feedback. This is the first Tree-sitter grammar I've created, so I'm sure I'm missing some tricks.

I also wrote a large-ish blog post on how to make a Tree-sitter grammar for a subset of Djot: https://www.jonashietala.se/blog/2024/03/19/lets_create_a_tree-sitter_grammar/

**Comment by NoahTheDuke** [comment]
This is cool as hell!

**Reply by NoahTheDuke** [reply]
I opened a PR on the `nvim-treesitter` repo to support djot using your grammar: https://github.com/nvim-treesitter/nvim-treesitter/pull/6327

**Reply by treeman** [reply]
Cool! I see there are some comments there already, I guess we'll backport the changes to this repo later.

**Reply by NoahTheDuke** [reply]
Yeah, I'll open an issue here once I get it all figured it.

**Reply by rtbs-dev** [reply]
Ooh! Would you be willing to PR for the helix editor? They include a lot of TS grammar's by default, adding this one would be awesome! 🙏

**Reply by NoahTheDuke** [reply]
I don't use and don't know anything about the Helix editor so I'm the wrong person to ask, but adding this to nvim-treesitter has been pretty easy so I expect adding it to Helix would be equally easy.

**Reply by treeman** [reply]
I also know nothing about the Helix editor, but looking at their repo it indeed doeesn't seem difficult. A merge request needs to:

1. Add it to `languages.toml`: https://github.com/helix-editor/helix/blob/master/languages.toml
2. Insert the queries here: https://github.com/helix-editor/helix/tree/master/runtime/queries

Basically the same as for Neovim.

It's great to be enthusiastic, but the query files are still a bit immature, so I don't think we need to rush too much.

Still, if anyone wants to do it then feel free. I opened an issue for it on the grammar's repo to track it: https://github.com/treeman/tree-sitter-djot/issues/10

**Reply by bdarcus** [reply]
I see the nvim-tree-sitter folks merged it. Looks good on quick testing!

![image](https://github.com/jgm/djot/assets/1134/148b9ec1-ce59-4e6a-af5e-905689d8d001)

**Reply by xxxserxxx** [reply]
> I also know nothing about the Helix editor, but looking at their repo it indeed doeesn't seem difficult.

This doesn't answer the PR question, but I did write a blog entry about [adding a tree sitter to Helix](https://ser1.net/post/adding-a-tree-sitter-to-helix/) as a plain user. Things have moved a bit since, making it easier, and especially your tree sitter went swimmingly. In particular, helix now works in the user directory and all of the faffing about with permissions isn't necessary.

It's really only a few lines in the user's `languages.toml` and couple of commands on the command line -- assuming you have Rust & git and all that already installed:

1. Add this to your `~/.config/helix/languages.toml` file (create it if you don't already have one):
   ```
   [[language]]
   name="djot"
   file-types=["dj"]
   auto-format=true
   scope="text.dj"
   roots=[]

   [[grammar]]
   name="djot"
   source={git="https://github.com/treeman/tree-sitter-djot", rev="0e9a836ec47612ade15645fb1680adb549894a6c"}
   ```
2. Run `helix -g fetch`
3. Run `helix -g build`
4. Symlink the grammars directory s.t. Helix can find them: `ln -s ~/.config/helix/runtime/grammars/sources/djot/queries ~/.config/helix/runtime/queries/djot`
5. Run `helix --health languages` and you should see djot; :
   ```
   $ hx --health languages | rg '^(Language|djot)'
   Language     LSP          DAP          Formatter    Highlight    Textobject   Indent
   djot         None         None         None         ✓            ✓            ✓
   ```

I'm sure there's a way to define the source without hard-coding a commit revision, but you can't leave the `rev` out and I haven't spent any time trying to figure it out. Anyway, it worked beautifully for me.

**Comment by jgm** [comment]
Amazing!  I'm starting an Emacs major mode that uses this.

**Reply by zetashift** [reply]
Hi, thank you for Djot! I was wondering if you ever ended up making the `djot-ts-mode`?

**Reply by jgm** [reply]
I have a work in progress.  It's fairly basic.
https://gist.github.com/jgm/17e17d6730efc253078b23e12aefd3df

**Comment by vassudanagunta** [comment]
I'm curious how it performs against the [manually coded implementations](https://github.com/jgm/djot?tab=readme-ov-file#implementations). Have you done any comparisons?

Also, does it make sense to either add your Tree-sitter grammar, or C and WASM parsers generated from it, to the above list?

**Reply by treeman** [reply]
I haven't done any performance comparisons against other implementations. I've only noted that it's fast enough for decently large files for my use-case.

Maybe it would make sense to add it to the implementations? But we should keep in mind that this is a Tree-sitter grammar, and should not be relied on to for instance generate html.

**Reply by vassudanagunta** [reply]
You mean to say it should only be used to implement syntax highlighting or to support syntax-aware editors?

If so, is it because renderers should be built against djot's standard AST, not the syntax tree produced by this Tree-sitter grammar? Does djot even have a standard AST, across the Lua, Javascript, Rust, Go, etc implementations?

Or is it because Tree-sitter parsers produce concrete syntax trees, not abstract syntax trees, more suitable for syntax highlighting and editors, not so suitable for compilers, interpreters or renderers?

**Reply by treeman** [reply]
Uh, I realize I'm out of my depth here and I should've said that I've _heard_ that dedicated parsers are better suited, but I'm no expert on the trade-offs here. *hides*

> Does djot even have a standard AST, across the Lua, Javascript, Rust, Go, etc implementations?

I don't think so no, at least with [jotdown](https://github.com/hellux/jotdown) you don't have an AST, only a stream of iterators to work with.

**Reply by vassudanagunta** [reply]
Fair enough. I'm out of my depth too, wrt to Tree-sitter, hence the questions. I'm working on a metalanguage called *Plain Text Style Sheets* which focuses on human oriented languages like Markdown. Parsers based on it would output a standard abstract grammar tree or semantic tree (I'm not sure what the right terms are).

**Comment by xxxserxxx** [comment]
The project README claims helix has syntax highlighting, which it doesn't. Helix, OOTB, does not know about djot. I think this should be clarified, at the very least with pointing to a snippet for how to add support to helix.

Here's what I have in my `~/.config/helix/languages.toml`:

```
[[grammar]]
name="djot"
source={git="https://github.com/treeman/tree-sitter-djot", rev="eb31845d59b9ee8c1b2098e78e9ca72004bd1579"}

[[language]]
name="djot"
file-types=["dj"]
auto-format=true
scope="text.dj"
soft-wrap.enable = true
roots=[]
grammar="djot"
```

**Reply by jgm** [reply]
I had thought someone had submitted a PR to helix but maybe it never happened.

Do you want to submit a PR modifying the README?

**Reply by rtbs-dev** [reply]
It looks like the PR was submitted and [merged into main](https://github.com/helix-editor/helix/pull/12562) on Feb 10. The latest release (25.01.1) is from Jan 25 though, so none of the installable versions have built-in support yet. The next release should include it though, based on the merge!

**Reply by xxxserxxx** [reply]
@rtbs-dev Oh, that's good. I run released, and didn't see it in the `--health` check.

**Comment by dereuromark** [comment]
I also spent quite some time refining syntax highlighting for Djot in various platforms, e.g. for github pages or blogs/websites:
https://php-collective.github.io/djot-php/cookbook/syntax-highlighting.html
The canonical sources are further down.

You can see it in action in the sandbox, or through the explained syntax (examples): https://php-collective.github.io/djot-php/guide/syntax.html

**Comment by dereuromark** [comment]
I decided to create a repo for grammars:
https://github.com/php-collective/djot-grammars

Feel free to PR yourself in the readme.
It's also reachable from the new awesome-djot list.
